### DISCUSSION TITLE
command line tools builds

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#256] [Design & Development]
Our current setup does not seem to have a mechanism to build any kind of command line tools that can be run from the unix shell from the command line.  At present, I think we have two that are not installed in a way the users can use them.  The ones I'm aware of are found in python/mspasspy/db/script.   They are command line tool called dbverify and dbclean.  

In the cxx tree we have the functionality of a "make install' that creates 3 directories under MSPASS_HOME (or it might be MSPASSHOME) called:  data, include, and lib.   I think we need:
1.  To add a "bin" directory to the install tree
2.  We need a mechanism to install python command line tools like dbclean and dbverify to that bin directory.

That is a suggestion, but the details about where this is installed and how are less important.  What we need is a standard place to put  command line tools that users would execute from the command line.  Since our container runs a unix os that is sensible and we could set up the container to have that directory in the path some a shell run from a notebook would immecdiately see those tools.   

This could be useful in a number of contexts:
1.   I can see uses for some simple command line tools that simplify certain standard operations like pretty printing of certain standard reports.   
2. I can see a reason for the evolution of adopting certain common C or Fortran programs for data prep.   Maybe less for the framework itself, but we should have a place for user's to put special stuff they need for their special applications using some legacy code.  

This came up because I encountered this issue when converting my plane wave migration code - an example legacy code.  I have a set of command line tools I use for data preparation that need a home and right now I don't know how to handle it short of a manual copy of the python scripts to some special directory.

**Comment by pavlis** [comment]
postscript to the previous.   I have a specific example script I want to generalize to put into mspass that has emerged as a need for handling the pwmig conversion.  The tool I have in mind is a generic tool to convert a table stored in a text file (csv or text file in fixed formats ala Antelope's Datascope) and store the results in a MongoDB collection.   I don't think the tool needs to be designed by committee, but I think you will likely agree when I'm done it will have some utility.  It needs to be a command line tool but in python, so resolving the topic of this discussion will be important.

**Comment by wangyinz** [comment]
The bin directory is already there. If you go to our mspass container, you can see that the two scripts (dbclean and dbverify) are in the correct directory:
```
# which mspass-dbclean
/usr/local/bin/mspass-dbclean
# ls /usr/local/bin/mspass*
/usr/local/bin/mspass-dbclean  /usr/local/bin/mspass-dbverify
```
Note that the setup will put them under the `/usr/local/bin` if we do a system install, and it should be under the `~/.local/` directory for the user install. I believe this is also the common behavior of Python packages. All you need to do is to create the script and then point to them in the `setup.py` script here: 
https://github.com/mspass-team/mspass/blob/35c5e77f8264b8bdd62dd291c38c77cbb19d3949/setup.py#L66-L71

To put a C binary into the same bin directory using the `setup.py` script is a bit tricky. I guess we could do something similar to what we setup for the ccore module here:
https://github.com/mspass-team/mspass/blob/35c5e77f8264b8bdd62dd291c38c77cbb19d3949/setup.py#L41-L43
All we need to do is set the [CMAKE_RUNTIME_OUTPUT_DIRECTORY](https://cmake.org/cmake/help/latest/variable/CMAKE_RUNTIME_OUTPUT_DIRECTORY.html) variable properly, but I am not exactly sure how at the moment. The other question is whether we have an absolute need to have binaries written in C for the main mspass package. It seems to me that this is better to be avoided.
